### PR TITLE
When available, use git rev-parse --show-toplevel to find root of project (#5588)

### PR DESCRIPTION
**Pull request by itsthejb**
Resolves https://github.com/tuist/tuist/issues/5588

### Short description 📝

* https://github.com/tuist/tuist/pull/5499 refactored `RootDirectoryLocating` to give precedence to `Workspace.swift` when locating the root of the project tree
* This breaks the ability to have nested workspaces, but also to shared code through ProjectDescriptionHelpers at the (true) root of the project
* The existing implementation relies entirely on inspecting the filesystem to work this out. However, it's possible to query Git directly to get the root of the local copy. In cases where this is possible (presence of Git doesn't need to be assumed), I can't think of any cases where this wouldn't be definitive

### How to test the changes locally 🧐

* An example project is provided in https://github.com/tuist/tuist/issues/5588
* Changes can be tested by running `tuist generate` for the nested workspace
* The "true" root of this project is then found and the shared ProjectDescriptionHelpers are compiled as expected

### Contributor checklist ✅

- [x] The code has been linted using run `make workspace/lint-fix`
  - Apparently unrelated errors:

```
warning: Skipping correcting file because it produced Swift parser errors: /Users/jc/Developer/OpenSource/tuist/Sources/tuist/tuist.docc/Tutorials/Tuist Tutorial/Getting-Started/Add External Dependencies/Dependencies.swift
{
  "diagnostics" : [
    "expected value in function call"
  ]
}
```

- [x] The change is tested via unit testing or acceptance testing, or both
  - Apparently unrelated errors:
  
```
TuistGenerateOneAcceptanceTests.xctest
GenerateOneAcceptanceTestInvalidWorkspaceManifestName
❌ TuistAcceptanceTesting/TuistAcceptanceTestCase.swift:31: Unexpectedly found nil while unwrapping an Optional value
2023-11-20 13:21:32.928967+0000 xctest[30443:9409074] TuistAcceptanceTesting/TuistAcceptanceTestCase.swift:31: Fatal error: Unexpectedly found nil while unwrapping an Optional value
objc[30445]: Class _TtC7Stencil7Context is implemented in both /Users/jc/Library/Developer/Xcode/DerivedData/Tuist-eiecjmhtbqlnyxcmnvdxvyeixocs/Build/Products/Debug/TuistScaffold.framework/Versions/A/TuistScaffold (0x10476eee0) and /Users/jc/Library/Developer/Xcode/DerivedData/Tuist-eiecjmhtbqlnyxcmnvdxvyeixocs/Build/Products/Debug/TuistGenerator.framework/Versions/A/TuistGenerator (0x105e943b8). One of the two will be used. Which one is undefined.
    ✖ Restarting after unexpected exit, crash, or test timeout in GenerateOneAcceptanceTestInvalidWorkspaceManifestName.test_invalid_workspace_manifest_name(); summary will include totals from previous launches.
Selected tests
TuistGenerateOneAcceptanceTests.xctest
GenerateOneAcceptanceTestInvalidWorkspaceManifestName
GenerateOneAcceptanceTestiOSAppWithFrameworks
❌ TuistAcceptanceTesting/TuistAcceptanceTestCase.swift:31: Unexpectedly found nil while unwrapping an Optional value
2023-11-20 13:21:33.380714+0000 xctest[30445:9409096] TuistAcceptanceTesting/TuistAcceptanceTestCase.swift:31: Fatal error: Unexpectedly found nil while unwrapping an Optional value
objc[30446]: Class _T
tC7Stencil7Context is implemented in both /Users/jc/Library/Developer/Xcode/DerivedData/Tuist-eiecjmhtbqlnyxcmnvdxvyeixocs/Build/Products/Debug/TuistScaffold.framework/Versions/A/TuistScaffold (0x1123b2ee0) and /Users/jc/Library/Developer/Xcode/DerivedData/Tuist-eiecjmhtbqlnyxcmnvdxvyeixocs/Build/Products/Debug/TuistGenerator.framework/Versions/A/TuistGenerator (0x113c043b8). One of the two will be used. Which one is undefined.
    ✖ Restarting after unexpected exit, crash, or test timeout in GenerateOneAcceptanceTestiOSAppWithFrameworks.test_ios_app_with_frameworks(); summary will include totals from previous launches.
Selected tests
TuistGenerateOneAcceptanceTests.xctest
GenerateOneAcceptanceTestiOSAppWithFrameworks
GenerateOneAcceptanceTestiOSAppWithHeaders
❌ TuistAcceptanceTesting/TuistAcceptanceTestCase.swift:31: Unexpectedly found nil while unwrapping an Optional value
2023-11-20 13:21:33.838657+0000 xctest[30446:9409120] TuistAcceptanceTesting/TuistAcceptanceTestCase.swift:31: Fatal error: Unexpectedly found nil while unwrapping an Optional value
objc[30448]: Class _TtC7Stencil7Context is implemented in both /Users/jc/Library/Developer/Xcode/DerivedData/Tuist-eiecjmhtbqlnyxcmnvdxvyeixocs/Build/Products/Debug/TuistScaffold.framework/Versions/A/TuistScaffold (0x1092beee0) and /Users/jc/Library/Developer/Xcode/DerivedData/Tuist-eiecjmhtbqlnyxcmnvdxvyeixocs/Build/Products/Debug/TuistGenerator.framework/Versions/A/TuistGenerator (0x10a9e43b8). One of the two will be used. Which one is undefined.
    ✖ Restarting after unexpected exit, crash, or test timeout in GenerateOneAcceptanceTestiOSAppWithHeaders.test_ios_app_with_headers(); summary will include totals from previous launches.
Selected tests
TuistGenerateOneAcceptanceTests.xctest
GenerateOneAcceptanceTestiOSAppWithHeaders
GenerateOneAcceptanceTestiOSAppWithTests
❌ TuistAcceptanceTesting/TuistAcceptanceTestCase.swift:31: Unexpectedly found nil while unwrapping an Optional value
2023-11-20 13:21:34.286731+0000 xctest[30448:9409130] TuistAcceptanceTesting/TuistAcceptanceTestCase.swift:31: Fatal error: Unexpectedly found nil while unwrapping an Optional value
objc[30449]: Class _TtC7Stencil7Context is implemented in both /Users/jc/Library/Developer/Xcode/DerivedData/Tuist-eiecjmhtbqlnyxcmnvdxvyeixocs/Build/Products/Debug/TuistScaffold.framework/Versions/A/TuistScaffold (0x104e76ee0) and /Users/jc/Library/Developer/Xcode/DerivedData/Tuist-eiecjmhtbqlnyxcmnvdxvyeixocs/Build/Products/Debug/TuistGenerator.framework/Versions/A/TuistGenerator (0x10659c3b8). One of the two will be used. Which one is undefined.
    ✖ Restarting after unexpected exit, crash, or test timeout in GenerateOneAcceptanceTestiOSAppWithTests.test_ios_app_with_tests(); summary will include totals from previous launches.
```
  
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated
  - It seems like a good idea to document that nested workspaces are currently only supported as part of a Git repository

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Tests/TuistCoreIntegrationTests/RootDirectoryLocatorIntegrationTests.swift [hunk 1]
```diff
@@ -154,4 +154,24 @@ final class RootDirectoryLocatorIntegrationTests: TuistTestCase {
         // Then
         XCTAssertEqual(got, temporaryDirectory)
     }
+
+    func test_locate_when_multiple_workspaces_exist_but_git_can_return_top_level() throws {
+        // Given
+        let temporaryDirectory = try AbsolutePath(validating: "/private")
+            .appending(temporaryPath().relative(to: "/"))
```
Comment (itsthejb) [inline Tests/TuistCoreIntegrationTests/RootDirectoryLocatorIntegrationTests.swift:161]: `/var` symlinks to `/private/var` on macOS. Temporary directory is made in `/var`, but Git will return absolute path from `/private/var`. So adjusting here seems like a simple solution

**Review by itsthejb — COMMENTED** [review]
Explanation

**Comment by pepicrft** [comment]
Hey @itsthejb 👋🏼 
Thanks for going ahead with this one. I'd instead [revert](https://github.com/tuist/tuist/pull/5637) the PR that introduced the regression in the first place and continue to use `Tuist/` as a reference of the root directory. It has worked fine, and I'd rather keep that logic as simple as possible.

**Comment by itsthejb** [comment]
@pepicrft Ok! Thanks for looking into it. Please feel free to close this one if you're going ahead with the other fix

**Pull request closed without merging**